File: payment-service/src/ps/models/audit.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class AuditTimelineStep(BaseModel):
    at: str | None = None
    step: str
    summary: str
    decision: str | None = None


class AuditPolicyExchange(BaseModel):
    """Provisional OPA exchange when no security event exists yet (e.g. chat preflight)."""

    evaluate_request: dict[str, Any] = Field(default_factory=dict)
    evaluate_response: dict[str, Any] = Field(default_factory=dict)


class AuditGovernance(BaseModel):
    security_event_id: str | None = None
    policy_exchange: AuditPolicyExchange | None = None


class CreateAuditExecutionRequest(BaseModel):
    capability: str = Field(default="CREATE_PAYMENT", min_length=1)
    skill: str = Field(default="create_payment", min_length=1)
    channel: str = Field(default="chat", min_length=1)
    status: str = Field(min_length=1)
    outcome: str = Field(min_length=1)
    request: dict[str, Any] = Field(default_factory=dict)
    interpretation: dict[str, Any] = Field(default_factory=dict)
    timeline: list[AuditTimelineStep] = Field(default_factory=list)
    timings_ms: dict[str, Any] = Field(default_factory=dict)
    result: dict[str, Any] = Field(default_factory=dict)
    governance: AuditGovernance | None = None
    execution_id: str | None = None


class PatchAuditExecutionRequest(BaseModel):
    status: str | None = None
    outcome: str | None = None
    timeline: list[AuditTimelineStep] | None = None
    timings_ms: dict[str, Any] | None = None
    result: dict[str, Any] | None = None
    governance: AuditGovernance | None = None

File: audit-service/src/audit_console/static/audit.js
const executionId = decodeURIComponent(location.pathname.split("/").filter(Boolean).slice(1).join("/"));

function field(label, value) {
  return `<div class="field"><dt>${label}</dt><dd class="mono">${value ?? "—"}</dd></div>`;
}

async function load() {
  document.getElementById("subtitle").textContent = executionId;
  const response = await AuditorAuth.apiFetch(`/api/audit-executions/${encodeURIComponent(executionId)}`);
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  const { execution } = await response.json();
  const actor = execution.actor || {};
  const request = execution.request || {};
  const result = execution.result || {};
  const governance = execution.governance || {};
  const timeline = (execution.timeline || []).map((step) =>
    `<li><strong>${step.step || "step"}${step.decision ? ` · ${step.decision}` : ""}</strong><br>${step.summary || ""}<br><span class="muted mono">${step.at || ""}</span></li>`
  ).join("");
  document.getElementById("summary").innerHTML = `
    <section class="card"><div class="card-header"><h2>${execution.capability || "Audit execution"}</h2><span class="outcome-${execution.outcome}">${execution.outcome || "—"}</span></div>
      <dl class="grid">
        ${field("Status", execution.status)}${field("Actor", actor.user_id)}
        ${field("Roles", (actor.roles || []).join(", "))}${field("Instruction", request.instruction_id)}
        ${field("Amount", request.amount)}${field("Value date", request.value_date)}
        ${field("Payment", result.payment_id)}${field("Security event", governance.security_event_id)}
      </dl>
    </section>
    <section class="card"><div class="card-header"><h2>Timeline</h2></div><ol class="timeline">${timeline || "<li>No steps recorded</li>"}</ol></section>`;
  document.getElementById("json").textContent = JSON.stringify(execution, null, 2);
}

async function loadOpa() {
  const response = await AuditorAuth.apiFetch(`/api/audit-executions/${encodeURIComponent(executionId)}/opa`);
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  const evidence = await response.json();
  document.getElementById("opa-source").textContent =
    evidence.source === "security_event"
      ? `Source: payment security event ${evidence.security_event_id}`
      : evidence.source === "policy_exchange"
        ? "Source: provisional preflight exchange"
        : "No OPA evidence linked";
  document.getElementById("opa-request").textContent = AuditorFormat.prettyJson(
    evidence.evaluate_request || {}
  );
  document.getElementById("opa-response").textContent = AuditorFormat.prettyJson(
    evidence.evaluate_response || {}
  );
  document.getElementById("opa-panel").classList.remove("hidden");
  if (evidence.security_event) {
    document.getElementById("event-json").textContent = AuditorFormat.prettyJson(
      evidence.security_event
    );
    document.getElementById("event-details").classList.remove("hidden");
  }
}

document.getElementById("load-opa").addEventListener("click", () =>
  void loadOpa().catch((error) => {
    document.getElementById("opa-source").textContent = `Load failed: ${error.message}`;
  })
);
AuditorAuth.bind(() => void load().catch((error) => {
  document.getElementById("summary").innerHTML = `<p class="empty">Load failed: ${error.message}</p>`;
}));
